airflow/contrib/hooks/pagerduty_hook.py

```diff
@@ -0,0 +1,130 @@
+            if password:
+                self.token = password
+
+            routing_key = conn.extra_dejson.get("routing_key")
+            if routing_key:
+                self.routing_key = routing_key
+
+        if token is not None:
+            self.token = token
+
+        if self.token is None:
+            raise AirflowException(
+                'Cannot get token: No valid api token nor pagerduty_conn_id supplied.')
+
+    # pylint: disable=too-many-arguments
+    def create_event(
+        self,
+        summary: str,
+        severity: str,
+        source: str = 'airflow',
+        action: str = 'trigger',
+        routing_key: Optional[str] = None,
+        dedup_key: Optional[str] = None,
+        custom_details: Optional[str] = None,
+        group: Optional[str] = None,
+        component: Optional[str] = None,
+        class_type: Optional[str] = None,
+        links: Optional[List[Dict]] = None,
+    ) -> Dict:
+        """
+        Create event for service integration.
+
+        :param summary: Summary for the event
+        :type summary: str
+        :param severity: Severity for the event, needs to be one of: Info, Warning, Error, Critical
+        :type severity: str
+        :param source: Specific human-readable unique identifier, such as a
+            hostname, for the system having the problem.
+        :type source: str
+        :param action: Event action, needs to be one of: trigger, acknowledge,
+            resolve. Default to trigger if not specified.
+        :type action: str
+        :param routing_key: Integration key. If not specified, will try to read
+            from connection's extra json blob.
+        :type routing_key: str
+        :param dedup_key: A string which identifies the alert triggered for the given event
+        :type dedup_key: str
+        :param custom_details: Free-form details from the event
+        :type custom_details: str
+        :param group: A cluster or grouping of sources. For example, sources
+            “prod-datapipe-02” and “prod-datapipe-03” might both be part of “prod-datapipe”
+        :type group: str
+        :param component: The part or component of the affected system that is broken.
+        :type component: str
+        :param class_type: The class/type of the event.
+        :type class_type: str
+        :param links: List of links to include.
+        :type class_type: list of str
+        """
+        if routing_key is None:
+            routing_key = self.routing_key
+        if routing_key is None:
+            raise AirflowException('No routing/integration key specified.')
+        payload = {
+            "summary": summary,
+            "severity": severity,
+            "source": source,
+        }
+        data = {
+            "routing_key": routing_key,
+            "event_action": action,
+            "payload": payload,
+        }
+        if custom_details:
+            payload["custom_details"] = custom_details
```

Comment:
Is it problematic if `payload` and `data` contain `None` values? If not you could perhaps build them by using dict literals and assigning values like `payload["component"] = component or None`.

Reply:
It will cause issue. Even if it doesn't, I would recommend against `foo or None` pattern if possible. It doesn't support proper `None` check and has edge-cases where values like `0` or `''` won't work. I have fixed a bug in k8s executor that's caused by exactly this issue :)

Reply:
Interesting. I did not know that thanks for letting me know. It's quite a shame since it's a very readable pattern in my opinion. 😅